definitions/marts/looker_studio_marts/ecf/ls_ecf_retention_kpi.sqlx

```diff
@@ -0,0 +1,172 @@
+WITH lead_providers AS (
+  SELECT
+     DISTINCT cpd_lead_provider_name AS lead_provider_name
+    ,participant_type
+  FROM
+    ${ref('ecf_retention_kpi_cut_off_date')}
+),
+
+--- Gather 2022 cut-off started & retained counts for each LP-PT combination.
+retention_2022_cut_off AS(
+  SELECT
+     cpd_lead_provider_name AS lead_provider_name
+    ,participant_type
+    ,COUNT(1) AS started_participants_2022_cut_off
+    ,SUM(CASE WHEN participant_retained_at_cut_off = TRUE THEN 1 ELSE 0 END) AS retained_participants_2022_cut_off
+  FROM
+    ${ref('ecf_retention_kpi_cut_off_date')}
+  WHERE
+    statement_cohort = 2022
+  GROUP BY
+     cpd_lead_provider_name
+    ,participant_type
+),
+
+--- Gather 2023 mid-point started & retained counts for each LP-PT combination.
+retention_2023_mid_point AS(
+  SELECT
+     cpd_lead_provider_name AS lead_provider_name
+    ,participant_type
+    ,COUNT(1) AS started_participants_2023_mid_point
+    ,SUM(CASE WHEN participant_retained_at_mid_point = TRUE THEN 1 ELSE 0 END) AS retained_participants_2023_mid_point
+  FROM
+    ${ref('ecf_retention_kpi_mid_point_date')}
+  WHERE
+    statement_cohort = 2023
+  GROUP BY
+     cpd_lead_provider_name
+    ,participant_type
+),
+
+--- Gather 2023 cut-off started & retained counts for each LP-PT combination.
+retention_2023_cut_off AS(
+  SELECT
+     cpd_lead_provider_name AS lead_provider_name
+    ,participant_type
+    ,COUNT(1) AS started_participants_2023_cut_off
+    ,SUM(CASE WHEN participant_retained_at_cut_off = TRUE THEN 1 ELSE 0 END) AS retained_participants_2023_cut_off
+  FROM
+    ${ref('ecf_retention_kpi_cut_off_date')}
+  WHERE
+    statement_cohort = 2023
+  GROUP BY
+     cpd_lead_provider_name
+    ,participant_type
+),
+
+--- Gather 2024 mid-point started & retained counts for each LP-PT combination.
+retention_2024_mid_point AS(
+  SELECT
+     cpd_lead_provider_name AS lead_provider_name
+    ,participant_type
+    ,COUNT(1) AS started_participants_2024_mid_point
+    ,SUM(CASE WHEN participant_retained_at_mid_point = TRUE THEN 1 ELSE 0 END) AS retained_participants_2024_mid_point
+  FROM
+    ${ref('ecf_retention_kpi_mid_point_date')}
+  WHERE
+    statement_cohort = 2024
+  GROUP BY
+     cpd_lead_provider_name
+    ,participant_type
+),
+
+--- Gather 2024 cut-off started & retained counts for each LP-PT combination.
+retention_2024_cut_off AS(
+  SELECT
+     cpd_lead_provider_name AS lead_provider_name
+    ,participant_type
+    ,COUNT(1) AS started_participants_2024_cut_off
+    ,SUM(CASE WHEN participant_retained_at_cut_off = TRUE THEN 1 ELSE 0 END) AS retained_participants_2024_cut_off
+  FROM
+    ${ref('ecf_retention_kpi_cut_off_date')}
+  WHERE
+    statement_cohort = 2024
+  GROUP BY
+     cpd_lead_provider_name
+    ,participant_type
+)
+
+--- Combine CTEs into final output where each point in time combination of started & retained appear as separate columns.

```

Comment:
So all of this is possible with a pivot. Simply do a CTE with a UNION ALL across group by's for cohort (one for mid_point, one for cut-off), make sure there is a column that contains the field you'll want as columns later (so a field with 'started_participants_2022_cut_off' etc) and then pivot into the format below. No joins necessary.